File: ios/SharedDataManager.m
// swift 개발환경인 경우 .swift 클래스에 코드 구현부 작성 후 .m 클래스를 통해 Objective-C인 ReactNative 환경에서 사용할 수 있도록 연결해주어야 한다.
#import <React/RCTBridgeModule.h>

@interface RCT_EXTERN_MODULE(SharedDataManager, NSObject)

RCT_EXTERN_METHOD(saveTodosToSharedDefaults:(NSString *)todosJsonString
               resolver:(RCTPromiseResolveBlock)resolve
               rejecter:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(getTodosFromSharedDefaults
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject)

@end

File: ios/SharedDataManager.swift
// SharedDataManager.swift
import Foundation
import WidgetKit

@objc(SharedDataManager)
class SharedDataManager: NSObject {

    static let appGroupId = "group.react.native.widget.todo"
    static let todoListKey = "todoListKey"

    @objc func saveTodosToSharedDefaults(_ todosJsonString: String, resolver resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
            if let sharedDefaults = UserDefaults(suiteName: Self.appGroupId) {
                sharedDefaults.set(todosJsonString, forKey: Self.todoListKey)
                if #available(iOS 15.0, *) {
                    WidgetCenter.shared.reloadAllTimelines()
                }
                resolve(nil)
            } else {
                let error = NSError(domain: "SharedDataManager", code: 500, userInfo: [NSLocalizedDescriptionKey: "Failed to access shared UserDefaults for App Group: \(Self.appGroupId)"])
                reject("shared_defaults_error", error.localizedDescription, error)
            }
        }


    @objc func getTodosFromSharedDefaults(_ resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
            if let sharedDefaults = UserDefaults(suiteName: Self.appGroupId) {
                let todosJsonString = sharedDefaults.string(forKey: Self.todoListKey)
                resolve(todosJsonString)
            } else {
                let error = NSError(domain: "SharedDataManager", code: 500, userInfo: [NSLocalizedDescriptionKey: "Failed to access shared UserDefaults for App Group: \(Self.appGroupId)"])
                reject("shared_defaults_error", error.localizedDescription, error)
            }
        }
}
